### PR TITLE
nixos/steam: add protontricks submodule

### PR DESCRIPTION
**Pull request by diniamo**
## Description of changes

The title is self-explanatory, as for the reason: other programs also rely on this variable to find proton versions, for example protontricks.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Shawn8901** [comment]
FYI in the PR adding that it was explicitly asked to not expose that globally: see https://github.com/NixOS/nixpkgs/pull/189398#discussion_r1437993188

**Comment by Atemu** [comment]
I don't think there's all that many tools which require steam-specific things. I'd be in favour of adding options for the one or two tools which actually need this. I honestly can't think of much else than `protontricks`.

As an escape hatch, you can also always run the tool inside steam-run.

**Comment by diniamo** [comment]
Tested, how is this?

File: nixos/modules/programs/steam.nix
Comment (Atemu) [inline nixos/modules/programs/steam.nix:151]: This does not need to be a submodule.

```suggestion
    protontricks = {
```

File: nixos/modules/programs/steam.nix
Comment (Atemu) [inline nixos/modules/programs/steam.nix:153]:
```suggestion
          package = mkPackageOption pkgs "protontricks";
```

File: nixos/modules/programs/steam.nix
Comment (Atemu) [inline nixos/modules/programs/steam.nix:160]: Move this to the implementation. `apply` is for when you to do some change so that it's visible to Nix code reading the same config option.

**Review by Atemu — CHANGES_REQUESTED** [review]

File: nixos/modules/programs/steam.nix
Comment (diniamo) [reply nixos/modules/programs/steam.nix:151]: Why? What's the point of having gamescopeSession as a submodule then?

File: nixos/modules/programs/steam.nix
Comment (diniamo) [reply nixos/modules/programs/steam.nix:160]: I'm not sure I understand, could you maybe give me an example?

File: nixos/modules/programs/steam.nix
Comment (Atemu) [reply nixos/modules/programs/steam.nix:151]: I don't see a reason for it to be a submodule either.

cc @nrdxp

File: nixos/modules/programs/steam.nix
Comment (diniamo) [reply nixos/modules/programs/steam.nix:160]: Then again, what's the reason it's used for the steam package?

Comment (Atemu) [reply nixos/modules/programs/steam.nix:160]: For it I could understand it slightly more as I could imagine usage by users outside of the steam module but it too should probably be moved to the implementation.

Note that this is just a minor issue.

Comment (diniamo) [reply nixos/modules/programs/steam.nix:160]: I see, so it's only if you want to expose it for external usage.

That's all my questions then, I'll try to implement the suggestions tomorrow.

**Comment by Atemu** [comment]
I'll take a closer look later but I don't see any obvious formatting issues. Unless you're doing something utterly weird that doesn't fit into the rest of Nixpkgs, there aren't any strict rules yet until https://github.com/NixOS/rfcs/pull/166 is implemented.

**Review by Shawn8901 — APPROVED** [review]
lgtm

**Comment by diniamo** [comment]
What is that ofborg error and where is the request change that's blocking the merge?

**Comment by Atemu** [comment]
It's an internal error that someone who maintains ofBorg will have to take a look at and manually remove at some point. It doesn't block the merge though, that's just my review. It'll stop blocking it as soon as I approve after I've taken a proper look or another committer comes around and dismisses it.

**Comment by diniamo** [comment]
Resolved conflicts, tested.

**Comment by diniamo** [comment]
@Atemu have you had a chance to look? It would be great if we could get this in before the next unstable merge.

File: nixos/modules/programs/steam.nix
Comment (Atemu) [inline nixos/modules/programs/steam.nix:29]: It'd be better if protontricks provided a proper override option for this. `overrideAttrs` isn't really a stable interface.

File: nixos/modules/programs/steam.nix
Comment (Atemu) [inline nixos/modules/programs/steam.nix:151]: Please consider users who may not know what protontricks is. The option to enable it should offer at least a little information on it.

We could also pull it from the package's meta like this: 

```suggestion
      enable = lib.mkEnableOption "protontricks, ${cfg.protontricks.package.meta.description}";
```

**Review by Atemu — COMMENTED** [review]
Two minor things, otherwise LGTM.

File: nixos/modules/programs/steam.nix
Comment (diniamo) [reply nixos/modules/programs/steam.nix:29]: I'm not sure what you mean by that / how it should be implemented. Do you mean the derivation should contain an argument for extraSteamCompatPaths, and we should use override here instead?

Comment (Atemu) [reply nixos/modules/programs/steam.nix:29]: Yup that's precisely it. You'd basically define an API that's quite a bit more stable than overrideAttrs.

**Comment by diniamo** [comment]
Looks like your suggestion fails the documentation CI, apparently option descriptions can't rely on config.

**Comment by diniamo** [comment]
Welp, adding it manually is the best I can come up with.

**Comment by Atemu** [comment]
> Looks like your suggestion fails the documentation CI (and not just the CI, my system config errors too), apparently option descriptions can't rely on config/pkgs.

Ah, indeed it makes sense; we don't have access to `config` when defining `options`. You could probably dig it out of `options..package.default` but hard-coding like this is fine.

**Review by Atemu — COMMENTED** [review]
Diff LGTM and the NixOS part appears to be working but I'm not sure protontricks is as it's giving me:

```
protontricks (ERROR): Could not find configured Proton installation!
protontricks (ERROR): Active Proton installation could not be found automatically.
Proton installation could not be found!
```

Though I'm also getting the same error when I `steam-run` it which worked before?

Is it working on your end @diniamo?

**Review by Atemu — APPROVED** [review]

**Comment by diniamo** [comment]
Nice, thanks

**Pull request merged**
